src/WebJobs.Extensions.DurableTask/ContextImplementations/DurableActivityContext.cs

```diff
@@ -55,13 +55,30 @@ internal JToken GetInputAsJson()
         {
             if (this.serializedInput != null && this.parsedJsonInput == null)
             {
-                JArray array = JArray.Parse(this.serializedInput);
-                if (array?.Count != 1)
+                if (this.messageDataConverter.JsonSettings.TypeNameHandling == TypeNameHandling.All)
                 {
-                    throw new ArgumentException("The serialized input is expected to be a JSON array with one element.");
+                    var objectArray = this.messageDataConverter.Deserialize<object[]>(this.serializedInput);
+
+                    if (objectArray?.Length != 1)
+                    {
+                        throw new ArgumentException("The serialized input is expected to be a JSON array with one element.");
+                    }
+
+                    var token = JToken.Parse(this.messageDataConverter.Serialize(objectArray[0]));
```

Comment:
I couldn't find a method that allowed me to do this easier, but I'd like to know if there is an easier way.

Reply:
It makes me wonder if we should even use this method inside of `GetInput(Type destinationType)`. I realize we have to for the case where the destination type is string so that we support the out-of-proc case, but maybe we just do something like below for the other case: 

`return this.messageDataConvert.Deserialize(this.serializedInput, destinationType.GetArrayType())[0]`;

On the other hand, we probably want to fix this method too, as it is used in other places.